### ISSUE TITLE
Option to add description to snapshot

### ISSUE THREAD

**Issue by erlanger** [restic/restic#2376]
Output of `restic version`
--------------------------
restic 0.9.5 (v0.9.5-46-g604b18aa) compiled with go1.12.7 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
Ability to add a description to a snapshot (i.e. "after upgrading OS to version 5.4.2")

What are you trying to do?
--------------------------
When I list snapshots, I would like an optional description describing the event that
caused us to make the snapshot (i.e. after upgrade to version 3.4.5)

Did restic help you or made you happy in any way?
-------------------------------------------------
Very good things:
* Static binary (which can live with the backups)
* Ability to mount the snapshots
* Save a good amount of space by deduping

**Comment by rawtaz** [comment]
Just add a tag to the snapshot. See `restic help backup`, in particular the `-t tag` argument. It's not free-form text, but good enough to do what you need if you just don't write such plain english. An example tag could be `postUpgradeOsToV542`.

**Comment by erlanger** [comment]
I knew about tags, but sometimes I want to include 'bullet points' in my description, such as:

"After running tronscript 10.8.2:
   * with mb disabled
   * and cleaners disabled"

I find it improper and difficult to put this in a tag.

**Comment by rawtaz** [comment]
Heh, seems terribly overkill to me. I can't even relate to wanting to do that. But I guess everyone has their use cases :-)

**Comment by moritzdietz** [comment]
I can see the usecase though. It's like a btrfs snapshot where you can add a comment or when you do a snapshot in VMWare; you can add a comment to it. I just checked and Borg has a feature where you can add a comment to a snapshot. Not to say that just because borg has it, restic must have it also, but just to highlight that there can be usecases.

I could see that one would do a `restic snapshots` and maybe filter for specific words in a comment field to find the right snapshot. But then again this could also be solved via tags - but maybe this is abusing the tag feature for something it shouldn't do.

**Comment by AlBundy33** [comment]
because restic uses principles from git i would say that the request sounds similar to a commit-message

**Comment by rehanone** [comment]
I agree with @moritzdietz. Like a comment on a git commit, a snapshot must have a description to make sense if you plan to keep them for years. In my opinion it a very serious design oversight which makes it more manifest the longer you use `restic` snapshots. I have been using it for 3 years now and cannot make any head and tail from the old snapshots I took in the beginning. I wish I had the ability to put comments on them to explain to my future self what are the reasons for those snapshots. This is a very serious issue in hindsight.

The way I think of it is somehow being able to add/link description to a snapshot like an annotated tag in git with the ability to add/edit/remove them from snapshots.

It would also make sense to have an option to exclude snapshots with descriptions from forget command if desired or marking a snapshot with a flag `protected` to indicate that it must be retained and excluded during forget operations.

**Comment by fabionebbia** [comment]
This would be very much appreciated

**Comment by ossie-git** [comment]
Started using Restic recently and was surprised that despite it being so feature-rich, it didn't have this basic feature. Would greatly appreciate adding it. Thanks

**Comment by Ambient-Impact** [comment]
Same. I've been trying to use a generated file list via `--files-from-raw` so I could have greater control over what gets stored but it's resulting in a comically huge message that's actually causing multiple GUIs ([restic-browser](https://github.com/emuell/restic-browser) and [backrest](https://github.com/garethgeorge/backrest)) to either freeze or fail to browse the snapshot contents; that's not restic's fault of course, but it does illustrate how unexpected this is.

**Comment by keyhelp** [comment]
I would also like to see this feature being added to Restic some day :+1:

**Comment by perariontaeadastra** [comment]
+1

**Comment by fabionebbia** [comment]
Thank you so much for adding this to your roadmap! ✨

**Comment by foxydevloper** [comment]
I'd love this! I started using restic because I wanted a tool exactly like it and it fit my use case perfectly! I think adding descriptions would be super useful since I do my backups manually, like if I have a folder I want to archive I'd manually run a backup and delete the folder. But the name of the folder alone isn't good enough to know what it was or why I archived it and the context.
This would fit a similar use case to https://github.com/restic/restic/issues/4026

**Comment by rafamadriz** [comment]
Just for the record [rustic](https://github.com/rustic-rs/rustic), a rust implementation of restic does have descriptions:

These are the options when you create a snapshot:
```
--description <DESCRIPTION>
   Add description to snapshot

--description-from <FILE>
   Add description to snapshot from file
```

And you can see the [comparison in the website](https://rustic.cli.rs/docs/comparison-restic.html?highlight=description#information-saved-in-snapshots), it can also save more information like label and delete protection.

Here's the related PR https://github.com/rustic-rs/rustic/pull/295

**Comment by MichaelEischer** [comment]
If someone wants to go ahead and implement support for a description field, just add a comment to this issue. It will still take a while until I get around to it.

**Comment by Laborratte5** [comment]
I would like to work on this issue.
My current plan would be to add a `--description` option to the `backup` command.
As I'm still learning the internals of restic I don't know how long it will take me to add a `description` command similar to the `tag` command for adding/editing/removing the description of existing snapshots.
Let me know if this would be helpful.

And thank you for building restic, it is an awesome tool and I hope I can give something back with this contribution.

**Comment by Laborratte5** [comment]
I added some questions about the `description` command to the draft pull request and would appreciate some feedback and opinions.

**Comment by Wonderer0** [comment]
@Laborratte5 Thanks for implementing this. As a newbie trying out Restic it's something I miss.  I have added a comment to the pull request about having a `--description-file` option, which would be especially useful for multi-line descriptions.

In some backups systems when mounting a backup it's possible to view a list of backups with their descriptions, and then pick the one you want to mount based on the description.  With Restic you have to mount the whole repo and then pick the one you want to open from a long list of snapshot directories, with timestamps as their names and no other information about them.  Of course you can can do this along side a list of snapshots, but that is less convenient.

It would be very convenient if the one line descriptions could be displayed as comments in Dolphin. It would also be useful other properties of the backups could be shown in Dolphin, such as tags in `Tags` and the first path in `Path`.

That is a suggestion for a further enhancement, and if it's a big job should not delay releasing the important work you have done.

**Comment by JakeB-OnGithub** [comment]
Just adding my voice to this issue - one use I have in particular is named/described snapshots for websites of various sizes.

Something I do regularly is make multiple backups of a given site while iteratively making changes:

1. Back up of site prior to any changes
2. Back up of site after I upload new data, but do not reference it
3. Back up of site after I mass rename so that new data is live and old data is still present on the server
4. Back up of site after mass deleting old data

Or something along those lines. If it turns out there was a problem in step 3, restic makes it easy for me to restore that - but in a month from now, I would only see 4x snapshots from a given date and no real information about which is which. I have actually "solved" the issue myself by adding a "Description.txt" to the project files that gets backed up with the project so I know what I did and when.

The easy alternative here is to just use git, and possibly git-lfs, which allows for commit messages, but it has its own issues. In the case above, if the site is quite small, git works well, but if the site has a large amount of media uploaded to it, I would prefer pulling a backup via restic.
